### ISSUE TITLE
Recurring Referrals Integration for 'WC Auto-Ship'

### ISSUE THREAD

**Issue by clickstudio** [affiliatewp/AffiliateWP#719]
I would like to add an integration for 'WC Auto-Ship' to your recurring referrals add on.
https://wooautoship.com/

It should be very straight forward using the following hook which fires after a successful order has been created from a schedule.

do_action( 'wc_autoship_payment_complete', $order_id, $schedule_id );

I have emailed Jon (developer) the link to this thread so he will be able to confirm this and add his thoughts.

**Comment by jonrules** [comment]
Hi AffiliateWP team,

I'm the developer for WC Auto-Ship. I've had several requests for integration with your WooCommerce software. It would be great if we could set up a call to discuss possible options. WC Auto-Ship is a plugin that allows customers to create flexible recurring order schedules in their WooCommerce account. It may be as easy as hooking into the wc_autoship_payment_complete hook, as clickstudio suggested.

**Comment by clickstudio** [comment]
Thanks Jon,

I'll test the mentioned hook ASAP and give some feedback.

**Comment by pippinsplugins** [comment]
Does it require Woo Subscriptions or is it independent?

**Comment by clickstudio** [comment]
It is completely independent Pippin.

**Comment by clickstudio** [comment]
What is the best AffiliateWP action to pass the order ID to that will ensure referral is processed ?

**Comment by clickstudio** [comment]
I haven't tested this but it should be as simple as this I reckon. if you have a better suggestion, please let me know.


add_action( 'wc_autoship_payment_complete', 'wc_autoship_to_affiliatewp' );

function wc_autoship_to_affiliatewp($order_id) {
	Affiliate_WP_WooCommerce::add_pending_referral( $order_id, '' );
	Affiliate_WP_WooCommerce::mark_referral_complete( $order_id  );
}

**Comment by ryanduff** [comment]
Doesn't the included WooCommerce integration work out of the box for this? I have a client that uses it and as far as I know, they're good to go. If anything, I think you just need the Lifetime Commissions add-on so that the subsequent auto-ship orders are properly credited.

**Comment by clickstudio** [comment]
I have lifetime subscriptions active and I am 99% sure I ran tests and it didn't work.
I'll test it again and report back.

**Comment by jonrules** [comment]
WC Auto-Ship should work with the Lifetime Commissions add-on if it is just looking for existing orders in WooCommerce. Autoship orders are created in the standard WooCommerce format. The customer's info (including email address) is attached to the order.

**Comment by ryanduff** [comment]
@jonrules 

Correct, that's what I alluded to last night. If the lifetime commissions extension is enabled for AffiliateWP, it should see the email address or user account the order is tied to and generate a referral. 

One caveat I just remembered from my testing is that the lifetime commission plugin doesn't set the lifetime affiliate info until the [first] order is marked complete. Once that's complete, the info should exist in the database for use on subsequent orders. So if you're testing and make an order and then a second order happens before the first order is marked complete, the second referral won't get generated. 

Maybe that was the issue OP was running into when testing.

**Comment by pippinsplugins** [comment]
Is there a reason the original order would be left as incomplete when a follow up order is processed?

**Comment by ryanduff** [comment]
@pippinsplugins Not generally in production. It was just something I noticed in testing. Nobody thinks to complete the first order before continuing with tests ;)

**Comment by clickstudio** [comment]
Good advice Ryan. I'm certain I never completed the orders during testing.
I'll go back and try again.

Cheers

**Comment by clickstudio** [comment]
Is there a particular reason why lifetime commissions requires the order to be completed before setting the affiliate info ?

There's a real chance that customers could create a second order before the first has been completed.

The 'Processing' order status is set once an order is fully paid so IMO it should be set at that time.
Wouldn't it also make sense to set the lifetime affiliate info prior to any order being placed ?

**Comment by pippinsplugins** [comment]
The order has to be completed because otherwise it could result in referrals being generated even though the original, referring order is never completed.

**Comment by clickstudio** [comment]
Thanks Pippin.

I guess from my perspective, once an order's status is 'Processing', it technically is completed from a payment perspective and this seems to be the way standard (original/non lifetime) referrals are handled.

Standard (original/non lifetime) referrals are been generated immediately after the order is paid (status is still processing), yet lifetime referrals requires those orders to be completed. Seems to be double standards in some respects.

**Comment by pippinsplugins** [comment]
Hmm, I think you might be right there. @sumobi could you confirm?

**Comment by sumobi** [comment]
Yeah that sounds fair to me, I'll jot this down for a future release.

**Comment by clickstudio** [comment]
@sumobi So in a future release lifetime referrals will be generated when the original order has a status of 'Processing' ?

What were your thoughts on setting the lifetime affiliate when the user hits the site rather than after an order ?

I have been testing things today and can't seem to get the referrals to generate after an Auto-Ship order. Can you tell me, if the original order has been deleted or is in trash would this affect anything ?

**Comment by sumobi** [comment]
@clickstudio 

> So in a future release lifetime referrals will be generated when the original order has a status of 'Processing' ?

I'm going to test the plugin and see if everything works properly on the processing hooks yes.

> What were your thoughts on setting the lifetime affiliate when the user hits the site rather than after an order ?

That's an interesting idea, we'll certainly consider that perhaps as an option.

> I have been testing things today and can't seem to get the referrals to generate after an Auto-Ship order. Can you tell me, if the original order has been deleted or is in trash would this affect anything ?

For Recurring Referrals or for Lifetime Commissions?

**Comment by clickstudio** [comment]
I am only using 'Lifetime Commissions' add-on, not 'Recurring Referrals'.

I have done some further tests using the following process:

- Lifetime Commissions add-on installed and enabled.

- Copied the referral link with affiliate ID
- Logged out and Pasted into browser.
- Placed an order with an auto-ship schedule to repeat in two days (created a new account on checkout).
- I then set the status of that original order to 'Completed'. The referral was generated as expected.

- The second auto-ship order was automatically generated today with no issues.
- I set the status of that order to 'Completed'.

No referral was generated......
